### PR TITLE
[wasm] Bump emscripten to 3.1.12

### PR DESCRIPTION
**Pull request by radekdoulik**

### CODE REVIEW

**Comment by radekdoulik** [comment]
@radical some tests timeout on windows, could you please take a look? like this one https://helix.dot.net/api/2019-06-17/jobs/fce081ba-bfdc-47e4-a4ae-5c8709e1cb11/workitems/normal-System.Composition.AttributeModel.Tests/console

could it be the python bug again?

**Comment by radical** [comment]
> could it be the python bug again?

No, python shouldn't be involved here. It's v8 timing out right before exit.

File: src/mono/wasm/wasm.proj [hunk 1]
```diff
@@ -137,6 +137,9 @@
       <_EmccLinkFlags Include="-s MODULARIZE=1"/>
       <_EmccLinkFlags Include="-Wl,--allow-undefined"/>
       <_EmccLinkFlags Include="-s ENVIRONMENT=&quot;web,webview,worker,node,shell&quot;" />
+      <!-- remove -Wno-limited-postlink-optimizations once below issue is fixed
+           active issue: https://github.com/emscripten-core/emscripten/pull/16727 -->
```
Comment (radical) [inline src/mono/wasm/wasm.proj:141]: The issue seems to be fixed. And the page says "first appeared in 3.1.11"

File: src/tests/BuildWasmApps/Wasm.Build.Tests/BlazorWasmBuildPublishTests.cs [hunk 2]
```diff
@@ -284,6 +284,9 @@ private string CreateProjectWithNativeReference(string id)
         {
             CreateBlazorWasmTemplateProject(id);
 
+            string extraProps = @"
+                <EmccExtraLDFlags>-sERROR_ON_UNDEFINED_SYMBOLS=0</EmccExtraLDFlags>
```
Comment (radical) [inline src/tests/BuildWasmApps/Wasm.Build.Tests/BlazorWasmBuildPublishTests.cs:288]: Is this still needed?

File: src/native/libs/CMakeLists.txt [hunk 3]
```diff
@@ -144,6 +144,7 @@ if (CLR_CMAKE_TARGET_UNIX OR CLR_CMAKE_TARGET_BROWSER)
         add_compile_options(-Wno-typedef-redefinition)
         add_compile_options(-Wno-c11-extensions)
         add_compile_options(-Wno-thread-safety-analysis)
+        add_compile_options(-Wno-strict-prototypes)
```
Comment (radical) [inline src/native/libs/CMakeLists.txt:147]: Just curious, why was this needed?

File: src/mono/wasm/wasm.proj (hunk 1 above)
Comment (radekdoulik) [reply src/mono/wasm/wasm.proj:141]: It is still happening in 3.1.12, for example here https://github.com/dotnet/runtime/runs/6869929890

File: src/tests/BuildWasmApps/Wasm.Build.Tests/BlazorWasmBuildPublishTests.cs (hunk 2 above)
Comment (radekdoulik) [reply src/tests/BuildWasmApps/Wasm.Build.Tests/BlazorWasmBuildPublishTests.cs:288]: I think you tested it recently and it was still needed, or was it the other test?

File: src/native/libs/CMakeLists.txt (hunk 3 above)
Comment (radekdoulik) [reply src/native/libs/CMakeLists.txt:147]: The compiler is more strict and there are many new warnings. I opened https://github.com/dotnet/runtime/issues/69142 so that we can fix it in another PR.

File: src/tests/BuildWasmApps/Wasm.Build.Tests/BlazorWasmBuildPublishTests.cs (hunk 2 above)
Comment (radekdoulik) [reply src/tests/BuildWasmApps/Wasm.Build.Tests/BlazorWasmBuildPublishTests.cs]: Nice, we can leave just the other workaround and remove this one.

**Comment by radical** [comment]
The WBT failures seem relevant here. The ones on `main` are with nodejs, but the ones failing here are with v8.

**Comment by radical** [comment]
wasm:
library test failure is https://github.com/dotnet/runtime/issues/71843
highresource_aot failures are https://github.com/dotnet/runtime/issues/71848

**Comment by radical** [comment]
wasm/HighResource_AOT failure is https://github.com/dotnet/runtime/issues/71848
The WBT failure is known, and unrelated to this PR.

**Review by radical — APPROVED** [review]
